shell/platform/android/android_context_gl.cc

```diff
@@ -105,10 +106,119 @@ static bool TeardownContext(EGLDisplay display, EGLContext context) {
   return true;
 }
 
+class AndroidEGLSurfaceDamage {
+ public:
+  void init(EGLDisplay display, EGLContext context) {
+    const char* extensions = eglQueryString(display, EGL_EXTENSIONS);
+
+    if (HasExtension(extensions, "EGL_KHR_partial_update")) {
+      set_damage_region = reinterpret_cast<PFNEGLSETDAMAGEREGIONKHRPROC>(
+          eglGetProcAddress("eglSetDamageRegionKHR"));
+    }
+
+    if (HasExtension(extensions, "EGL_EXT_swap_buffers_with_damage")) {
+      swap_buffers_with_damage =
+          reinterpret_cast<PFNEGLSWAPBUFFERSWITHDAMAGEEXTPROC>(
+              eglGetProcAddress("eglSwapBuffersWithDamageEXT"));
+    } else if (HasExtension(extensions, "EGL_KHR_swap_buffers_with_damage")) {
+      swap_buffers_with_damage =
+          reinterpret_cast<PFNEGLSWAPBUFFERSWITHDAMAGEEXTPROC>(
+              eglGetProcAddress("eglSwapBuffersWithDamageKHR"));
+    }
+
+    _partial_redraw_supported =
+        set_damage_region != nullptr && swap_buffers_with_damage != nullptr;
+  }
+
+  void SetDamageRegion(EGLDisplay display,
+                       EGLSurface surface,
+                       const std::optional<SkIRect>& region) {
+    if (set_damage_region && region) {
+      auto rects = RectToInts(display, surface, *region);
+      set_damage_region(display, surface, rects.data(), 1);
+    }
+  }
+
+  std::optional<SkIRect> InitialDamage(EGLDisplay display, EGLSurface surface) {
+    if (!_partial_redraw_supported) {
+      return std::nullopt;
+    }
+
+    EGLint age;
+    eglQuerySurface(display, surface, EGL_BUFFER_AGE_EXT, &age);
+
+    if (age == 0) {  // full repaint
+      return std::nullopt;
+    } else if (age == 1) {
+      // no initial damage
+      return SkIRect::MakeEmpty();
+    } else {
+      age -= 2;
```

Comment:
given that we are using `2` here and in `damage_history_.size() > 2`, consider moving it to a constant.